server/api/impl/admin/synchronize.go

```diff
@@ -180,7 +113,121 @@ func (adm *Admin) UninstallApp(appID apps.AppID) error {
 	return nil
 }
 
-func (adm *Admin) RegisterApp(manifest *apps.Manifest) error {
+func getAppsForRegistration(path string) (apps.AppVersionMap, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't read %s file", path)
+	}
+	var apps *AppVersions
+	if err := json.Unmarshal(data, &apps); err != nil || apps == nil {
+		return nil, errors.Wrapf(err, "can't unmarshal %s file", appsJSONFile)
+	}
+
+	return apps.Apps, nil
+}
+
+func (adm *Admin) populateManifests(appVersions apps.AppVersionMap) {
+	adm.store.Manifest().Cleanup()
+
+	for id, version := range appVersions {
+		manifest, err := adm.awsClient.GetManifest(id, version)
+		if err != nil {
+			// Note that we are not returning an error here. Instead we drop the app from the list.
+			delete(appVersions, id)
+
+			adm.mm.Log.Error("failed to get manifest", "app", id, "version", version, "err", err.Error())
+
+			continue
+		}
+
+		adm.store.Manifest().Save(manifest)
+	}
+}
+
+// mergeApp merges too files and detects apps that should upgrade, register or remove
+// appsForRegistration are apps from the apps.json file
+// registeredApps are apps in the config already registered.
+func mergeApps(appsForRegistration apps.AppVersionMap, registeredApps []*apps.App) (apps.AppVersionMap, map[apps.AppID]appNewVersion, map[apps.AppID]*apps.App) {
+	appsToRegister := apps.AppVersionMap{}
+	appsToUpgrade := map[apps.AppID]appNewVersion{}
+	appsToRemove := map[apps.AppID]*apps.App{}
+
+	allRegisteredAppsMap := map[apps.AppID]*apps.App{}
+	for _, registeredApp := range registeredApps {
+		allRegisteredAppsMap[registeredApp.AppID] = registeredApp
+		newVersion, ok := appsForRegistration[registeredApp.AppID]
+		if !ok {
+			// app was registered but it is not in the registration list anymore.
+			appsToRemove[registeredApp.AppID] = registeredApp
+			continue
+		}
+		if newVersion != registeredApp.Manifest.Version {
+			appsToUpgrade[registeredApp.AppID] = appNewVersion{registeredApp, newVersion}
+		}
+	}
+	for id, appToRegister := range appsForRegistration {
+		if _, ok := allRegisteredAppsMap[id]; !ok {
+			// here is a new app
+			appsToRegister[id] = appToRegister
+		}
+	}
+	return appsToRegister, appsToUpgrade, appsToRemove
+}
+
+func (adm *Admin) removeApps(appsToRemove map[apps.AppID]*apps.App) {
+	for _, appToRemove := range appsToRemove {
+		switch appToRemove.Status {
+		case apps.AppStatusInstalled:
+			if err := adm.UninstallApp(appToRemove.AppID); err != nil {
+				adm.mm.Log.Error("can't uninstall app", "app_id", appToRemove.AppID, "err", err.Error())
+			}
+		case apps.AppStatusRegistered:
+			// delete app from proxy plugin, not removing the data
+			if err := adm.store.App().Delete(appToRemove); err != nil {
+				adm.mm.Log.Error("can't delete app from store", "app_id", appToRemove.AppID, "err", err.Error())
+			}
+		}
+	}
+}
+
+func (adm *Admin) upgradeApps(appsToUpgrade map[apps.AppID]appNewVersion) map[apps.AppID]appOldVersion {
```

Comment:
somewhat counter-intuitive that the `upgradeApps` method takes `appNewVersion` objs as input and returns `appOldVersion` objs. I would've expected the opposite.